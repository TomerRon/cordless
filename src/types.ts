import { ApplicationCommandOptionAllowedChannelTypes } from '@discordjs/builders'
import { APIApplicationCommandOptionChoice } from 'discord-api-types/v10'
import Discord, {
  ApplicationCommandOptionType,
  ClientEvents,
  ClientOptions,
  CommandInteraction,
} from 'discord.js'

/** Initialization options for your cordless bot */
export type InitOptions<C extends CustomContext = {}> = {
  /** The commands used by your bot. */
  commands?: BotCommand<C>[]
  /** The functions used by your bot */
  functions: BotFunction<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
  /**
   * Your bot token.
   *
   * @see https://discordjs.guide/preparations/setting-up-a-bot-application.html#your-bot-s-token
   */
  token: string
  /** A custom context object which will extend the context passed to your commands and functions */
  context?: C
  /**
   * Override the default Gateway Intents of the discord.js client.
   *
   * By default, the discord.js client will initialize with the [GUILDS, GUILD_MESSAGES] intents.
   * These default intents should be sufficient in most cases.
   *
   * @see https://discord.com/developers/docs/topics/gateway#gateway-intents
   * @see https://discordjs.guide/popular-topics/intents.html
   */
  intents?: ClientOptions['intents']
  /**
   * Generate a help function for the bot, which will be triggered by the value.
   *
   * For example, given a value of "!help", the following commands will be available:
   *
   * - !help (shows a list of available functions)
   * - !help <function-name> (shows a description and usage instructions for the given function)
   */
  helpCommand?: string
}

export type BotCommand<C extends CustomContext = {}> =
  | BotCommandWithHandler<C>
  | BotCommandWithSubcommands<C>

export interface BotCommandWithHandler<C extends CustomContext = {}>
  extends BotCommandBase {
  handler: (args: BotCommandHandlerArgs<C>) => void | Promise<void>
  options?: BotCommandOption[]
  subcommands?: never
}

export interface BotCommandWithSubcommands<C extends CustomContext = {}>
  extends BotCommandBase {
  subcommands: BotCommandWithHandler<C>[]
  handler?: never
  options?: never
}

type BotCommandBase = {
  name: string
  description?: string
}

export type BotCommandHandlerArgs<C extends CustomContext = {}> = {
  interaction: CommandInteraction
  context: Context<C>
}

export type BotCommandOption =
  | BotCommandStringOption
  | BotCommandIntegerOption
  | BotCommandBooleanOption
  | BotCommandUserOption
  | BotCommandChannelOption
  | BotCommandRoleOption
  | BotCommandMentionableOption
  | BotCommandNumberOption
  | BotCommandAttachmentOption

export interface BotCommandStringOption extends BotCommandOptionBase {
  type: 'STRING'
  choices?: APIApplicationCommandOptionChoice<string>[]
}

export interface BotCommandIntegerOption extends BotCommandOptionBase {
  type: 'INTEGER'
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandBooleanOption extends BotCommandOptionBase {
  type: 'BOOLEAN'
}

export interface BotCommandUserOption extends BotCommandOptionBase {
  type: 'USER'
}

export interface BotCommandChannelOption extends BotCommandOptionBase {
  type: 'CHANNEL'
  channelTypes?: ApplicationCommandOptionAllowedChannelTypes[]
}

export interface BotCommandRoleOption extends BotCommandOptionBase {
  type: 'ROLE'
}

export interface BotCommandMentionableOption extends BotCommandOptionBase {
  type: 'MENTIONABLE'
}

export interface BotCommandNumberOption extends BotCommandOptionBase {
  type: 'NUMBER'
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandAttachmentOption extends BotCommandOptionBase {
  type: 'ATTACHMENT'
}

type BotCommandOptionBase = {
  type: ApplicationCommandOptionType
  name: string
  description?: string
  required?: boolean
}

export type BotFunction<
  E extends keyof ClientEvents = 'messageCreate',
  C extends CustomContext = {},
> = {
  /**
   * The name of this function (no spaces allowed).
   * It will be displayed in the help function if there is a helpCommand.
   */
  name?: string
  /**
   * The description of this function and usage instructions.
   * It will be displayed in the help function if there is a helpCommand.
   */
  description?: string
  /**
   * The event this function should subscribe to (default: "messageCreate").
   */
  event?: E
  /** Determines whether or not this function should run */
  condition: (...args: [...ClientEvents[E], Context<C>]) => boolean
  /** Called whenever this function should run */
  callback: (
    ...args: [...ClientEvents[E], Context<C>]
  ) => void | Promise<Discord.Message | void>
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomContext = Record<string, any>

export type Context<C extends CustomContext = {}> = {
  client: Discord.Client<true>
  functions: BotFunction<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
} & C
