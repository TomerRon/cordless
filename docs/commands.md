### Commands

Commands are the easiest way to let users interact with your bot. Cordless allow you to interface with the full [Discord Application Commands API](https://discord.com/developers/docs/interactions/application-commands) in a declarative fashion. You can enhance your commands with interactive buttons, modals, autocompletion, subcommands, and more.

#### Basic usage

You can declare a command by giving it a name and a handler:

```ts
// TypeScript
import { BotCommand } from 'cordless'

const ping: BotCommand = {
  name: 'ping',
  description: 'Reponds to your ping with a pong!', // optional
  handler: ({ interaction }) => interaction.reply('Pong!'),
}
```

```js
// JavaScript
const ping = {
  name: 'ping',
  description: 'Reponds to your ping with a pong!', // optional
  handler: ({ interaction }) => interaction.reply('Pong!'),
}
```

Register the command by passing it to the `init` method:

```ts
init({
  // ...
  commands: [ping],
  token: 'your.bot.token',
})
```

Your bot can now respond to the `/ping` command.

#### Components

You can add [Message Components](https://discord.com/developers/docs/interactions/message-components) to your interactions - these components include interactive buttons, link buttons, modals, select menus, and more.

For more information about using the components API, see: [docs/command-components.md](command-components.md)

```ts
const ping: BotCommand = {
  name: 'ping',
  components: [
    {
      label: 'Ping again',
      style: 'PRIMARY',
      handler: ({ interaction }) => interaction.reply('Pong again!'),
    },
  ],
  handler: ({ interaction, components }) =>
    interaction.reply({
      content: 'Pong!',
      components,
    }),
}
```

#### Options

You can allow your command to receive options (arguments) by adding an `options` array.

For more information about using the options API, see: [docs/command-options.md](command-options.md)

```ts
const commandWithOptions: BotCommand = {
  // ...
  options: [
    {
      type: 'INTEGER',
      name: 'num',
      required: true,
    },
    {
      type: 'STRING',
      name: 'language',
      choices: [
        {
          name: 'English',
          value: 'en',
        },
        {
          name: 'Deutsch',
          value: 'de',
        },
        {
          name: 'Fran√ßais',
          value: 'fr',
        },
      ],
    },
  ],
  // ...
}
```

#### Subcommands

[Subcommands](https://discord.com/developers/docs/interactions/application-commands#subcommands-and-subcommand-groups) organize your commands by specifying actions within a command or group.

You can create a command with subcommands by passing a `subcommands` array.

For more information about using the subcommands API, see: [docs/command-subcommands.md](command-subcommands.md)
