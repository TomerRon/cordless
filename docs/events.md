### Event Handlers and Gateway Events

Event handlers allow you to subscribe to any [Discord Gateway Event](https://discord.com/developers/docs/topics/gateway#commands-and-events-gateway-events), such as:

- A user joined/left the server
- A message was deleted
- A new channel was created
- etc...

For example, let's say our bot needs to greet new channels whenever they are created, expect for channels that start with `admin-`. We can subscribe an event handler to the `channelCreate` event:

```ts
// TypeScript
const channelGreeter: BotEventHandler<'channelCreate'> = {
  event: 'channelCreate',
  condition: (channel) => !channel.name.startsWith('admin-'),
  callback: (channel) => {
    if (channel.isText()) {
      return channel.send(`Hello world! This is ${channel.name}`)
    }
  },
}
```

```ts
// JavaScript
const channelGreeter = {
  event: 'channelCreate',
  condition: () => true,
  callback: async (channel) => {
    if (channel.isText()) {
      await channel.send(`Hello world! This is ${channel.name}`)
    }
  },
}
```

We can then add this event handler to our bot on initialization:

```ts
init({
  // ...
  handlers: [channelGreeter],
  token: 'your.bot.token',
})
```

#### Events and Intents

By default, cordless initializes the discord.js client with the [Gateway Intents](https://discord.com/developers/docs/topics/gateway#gateway-intents) `[GUILDS, GUILD_MESSAGES]`. When you start subscribing to events, you may need to specify additional intents in your bot.

For more information, see: [docs/intents.md](intents.md)

#### Gateway Events reference

For a full list of Gateway Events, see: https://discord.com/developers/docs/topics/gateway#commands-and-events-gateway-events
