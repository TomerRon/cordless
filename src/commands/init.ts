import { Client } from 'discord.js'
import { BotCommand, Context, CustomContext } from '../types'
import buildCommands from './builders/buildCommands'
import handleButton from './handlers/handleButton'
import handleCommand from './handlers/handleCommand'
import getButtonHandlerMap from './utils/getButtonHandlerMap'
import { registerCommands } from './utils/rest'

export type InitCommandsArgs<C extends CustomContext> = {
  client: Client<true>
  commands: BotCommand<C>[]
  context: Context<C>
  token: string
}

const initCommands = <C extends CustomContext>({
  client,
  commands,
  context,
  token,
}: InitCommandsArgs<C>) => {
  const resolvedCommands = buildCommands(commands)

  registerCommands({
    applicationId: client.application.id,
    commands: resolvedCommands,
    token,
  })

  if (!resolvedCommands.length) {
    return
  }

  const buttonHandlerMap = getButtonHandlerMap(commands)

  client.on('interactionCreate', (interaction) => {
    if (interaction.isCommand()) {
      return handleCommand({
        commands,
        context,
        interaction,
      })
    }

    if (interaction.isButton()) {
      return handleButton({ buttonHandlerMap, interaction, context })
    }
  })
}

export default initCommands
