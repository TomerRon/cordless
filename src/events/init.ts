import { Client } from 'discord.js'
import { BotFunction, Context, CustomContext } from '../types'
import handleEvent from '../utils/handleEvent'

export type InitEventsArgs<C extends CustomContext = {}> = {
  client: Client<true>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  functions: BotFunction<any, C>[]
  context: Context<C>
}

const initEvents = <C extends CustomContext>({
  client,
  functions,
  context,
}: InitEventsArgs<C>) => {
  const eventFunctionsMap = functions.reduce<
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    Record<string, BotFunction<any, C>[]>
  >((acc, curr) => {
    const key = curr.event || 'messageCreate'

    return {
      ...acc,
      [key]: [...(acc[key] || []), curr],
    }
  }, {})

  Object.entries(eventFunctionsMap).forEach(([event, eventFns]) => {
    client.on(event, (...args) => handleEvent(args, eventFns, context))
  })
}

export default initEvents
