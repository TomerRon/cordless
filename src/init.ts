import Discord from 'discord.js'
import { InitOptions } from './types'
import handleMessage from './utils/handleMessage'

/**
 * Initializes a cordless bot with the given options.
 * Returns a discord.js client.
 */
export const init = (options: InitOptions): Discord.Client => {
  const client = new Discord.Client()

  client.on('message', (msg) => handleMessage(msg, client, options))

  return client
}
