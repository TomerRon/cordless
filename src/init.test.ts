import { Client, Intents } from 'discord.js'
import * as initCommandsModule from './commands/init'
import * as initEventsModule from './events/init'
import { init } from './init'
import { BotCommand, BotEventHandler, InitOptions } from './types'

const onceSpy = jest.fn()
const loginSpy = jest.fn()
const mockNotLoggedInClient = {
  once: onceSpy,
  login: loginSpy,
} as unknown as Client<false>

const mockLoggedInClient = {
  application: {
    id: 'mock-application-id',
  },
} as unknown as Client<true>

onceSpy.mockImplementation((_, callback) => callback(mockLoggedInClient))

jest.mock('discord.js', () => {
  const originalModule = jest.requireActual('discord.js')

  return {
    ...originalModule,
    Client: jest.fn().mockImplementation(() => mockNotLoggedInClient),
  }
})

describe('init', () => {
  beforeEach(jest.clearAllMocks)

  const mockToken = 'mock-token'

  const mockCommands: BotCommand[] = [
    {
      name: 'ping',
      description: 'ping-desc',
      handler: jest.fn(),
    },
  ]

  const mockEventHandler: BotEventHandler = {
    condition: (msg) => msg.content === 'ping',
    callback: jest.fn(),
  }

  const mockOptions: InitOptions = {
    commands: mockCommands,
    handlers: [mockEventHandler],
    token: mockToken,
  }

  const initCommandsSpy = jest
    .spyOn(initCommandsModule, 'default')
    .mockReturnValue(undefined)

  const initEventsSpy = jest
    .spyOn(initEventsModule, 'default')
    .mockReturnValue(undefined)

  const setupTest = (options?: Partial<InitOptions>) =>
    init({ ...mockOptions, ...options })

  it('should initialize the client with the default intents', async () => {
    await setupTest()

    expect(Client).toHaveBeenCalledWith({
      intents: [Intents.FLAGS.GUILD_MESSAGES, Intents.FLAGS.GUILDS],
    })
  })

  it('should initialize the client with the provided intents', async () => {
    const intents = [
      Intents.FLAGS.GUILD_BANS,
      Intents.FLAGS.GUILD_EMOJIS_AND_STICKERS,
    ]

    await setupTest({ intents })

    expect(Client).toHaveBeenCalledWith({
      intents,
    })
  })

  it('should call initCommands with the given commands', async () => {
    await setupTest()

    expect(initCommandsSpy).toHaveBeenCalledWith({
      client: mockLoggedInClient,
      commands: mockCommands,
      context: {
        client: mockLoggedInClient,
        handlers: mockOptions.handlers,
      },
      token: mockToken,
    })
  })

  it('should call initCommands with an empty list of commands by default', async () => {
    await setupTest({ commands: undefined })

    expect(initCommandsSpy).toHaveBeenCalledWith({
      client: mockLoggedInClient,
      commands: [],
      context: {
        client: mockLoggedInClient,
        handlers: mockOptions.handlers,
      },
      token: mockToken,
    })
  })

  it('should call initEvents with the given event handlers', async () => {
    await setupTest()

    expect(initEventsSpy).toHaveBeenCalledWith({
      client: mockLoggedInClient,
      handlers: [mockEventHandler],
      context: {
        client: mockLoggedInClient,
        handlers: mockOptions.handlers,
      },
    })
  })

  it('should call initEvents with an empty list of event handlers by default', async () => {
    await setupTest({ handlers: undefined })

    expect(initEventsSpy).toHaveBeenCalledWith({
      client: mockLoggedInClient,
      handlers: [],
      context: {
        client: mockLoggedInClient,
        handlers: [],
      },
    })
  })

  it('should return a logged-in client', async () => {
    const client = await setupTest()

    expect(client).toStrictEqual(mockLoggedInClient)
    expect(loginSpy).toHaveBeenCalledWith(mockToken)
    expect(onceSpy).toHaveBeenCalledWith('ready', expect.any(Function))
  })
})
