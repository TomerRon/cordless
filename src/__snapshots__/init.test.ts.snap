// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init should handle many functions and map each one to an event handler 1`] = `
Array [
  Array [
    Array [
      Object {
        "content": "ping",
      },
    ],
    Array [
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "name": "Function1",
      },
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "messageCreate",
        "name": "Function3",
      },
    ],
    Object {
      "client": Object {
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "messageCreate",
              [Function],
            ],
            Array [
              "channelCreate",
              [Function],
            ],
            Array [
              "messageDelete",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "functions": Array [
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "name": "Function1",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "channelCreate",
          "name": "Function2",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "messageCreate",
          "name": "Function3",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "messageDelete",
          "name": "Function4",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "channelCreate",
          "name": "Function5",
        },
      ],
    },
  ],
  Array [
    Array [
      Object {
        "content": "ping",
      },
    ],
    Array [
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "channelCreate",
        "name": "Function2",
      },
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "channelCreate",
        "name": "Function5",
      },
    ],
    Object {
      "client": Object {
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "messageCreate",
              [Function],
            ],
            Array [
              "channelCreate",
              [Function],
            ],
            Array [
              "messageDelete",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "functions": Array [
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "name": "Function1",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "channelCreate",
          "name": "Function2",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "messageCreate",
          "name": "Function3",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "messageDelete",
          "name": "Function4",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "channelCreate",
          "name": "Function5",
        },
      ],
    },
  ],
  Array [
    Array [
      Object {
        "content": "ping",
      },
    ],
    Array [
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "messageDelete",
        "name": "Function4",
      },
    ],
    Object {
      "client": Object {
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "messageCreate",
              [Function],
            ],
            Array [
              "channelCreate",
              [Function],
            ],
            Array [
              "messageDelete",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "functions": Array [
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "name": "Function1",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "channelCreate",
          "name": "Function2",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "messageCreate",
          "name": "Function3",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "messageDelete",
          "name": "Function4",
        },
        Object {
          "callback": [MockFunction],
          "condition": [MockFunction],
          "event": "channelCreate",
          "name": "Function5",
        },
      ],
    },
  ],
]
`;
